jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/ThreadLimitHandlerTest.java

```diff
@@ -378,4 +386,127 @@ public void run()
 
         await().atMost(5, TimeUnit.SECONDS).until(handler::getRemoteCount, is(0));
     }
+
+    @Test
+    public void testBlockingWrite() throws Exception
+    {
+        ThreadLimitHandler handler = new ThreadLimitHandler("Forwarded");
+        handler.setThreadLimit(1);
+
+        CompletableFuture<Throwable> future = new CompletableFuture<>();
+        handler.setHandler(new Handler.Abstract()
+        {
+            @Override
+            public boolean handle(Request request, Response response, Callback callback)
+            {
+                try
+                {
+                    while (true)
+                    {
+                        try (Blocker.Callback blocking = Blocker.callback())
+                        {
+                            response.write(false, BufferUtil.toBuffer("x".repeat(1024)), blocking);
+                            blocking.block();
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    future.complete(e);
+                    callback.failed(e);
+                }
+                return true;
+            }
+        });
+        _server.setHandler(handler);
+        _server.start();
+
+        Socket client = new Socket("127.0.0.1", _connector.getLocalPort());
+        client.getOutputStream().write(("POST /" + " HTTP/1.0\r\nForwarded: for=1.2.3.4\r\nContent-Length: 0\r\n\r\n").getBytes());
+
+        // Validate the first 100 bytes of the response.
+        byte[] bytes = client.getInputStream().readNBytes(100);
+        String utf8String = StringUtil.toUTF8String(bytes, 0, bytes.length);
+        assertThat(utf8String, Matchers.containsString(" 200 OK"));
+        assertThat(utf8String, Matchers.containsString("xxxxx"));
+
+        // Delay to let the write side get TCP blocked, then close the connection.
+        // This ensures the server write callback will be failed by a different thread.
+        Thread.sleep(1000);
```

Comment:
Please do not rely on time. If you want to wait until the writer is blocked because of TCP congestion, use the following trick:

 - create an atomic ref: `AtomicReference<Thread> threadRef = new AtomicReference<>();`
 - fill the ref from the handler: `threadRef.set(Thread.currentThread());`
 - use awaitility to wait for the thread to be in WAITING state:
```java
        await().atMost(5, TimeUnit.SECONDS).until(() ->
        {
            Thread thread = threadRef.get();
            return thread != null && thread.getState() == Thread.State.WAITING;
        });
```

Reply:
This worked, thanks.